crates/polars-time/src/replace.rs

```diff
@@ -0,0 +1,156 @@
+use polars_core::prelude::*;
+
+use crate::prelude::*;
+
+/// Replace specific time component of a `DatetimeChunked` with a specified value.
+#[cfg(feature = "dtype-datetime")]
+#[allow(clippy::too_many_arguments)]
+pub fn replace_datetime(
+    ca: &DatetimeChunked,
+    year: &Int32Chunked,
+    month: &Int8Chunked,
+    day: &Int8Chunked,
+    hour: &Int8Chunked,
+    minute: &Int8Chunked,
+    second: &Int8Chunked,
+    microsecond: &Int32Chunked,
+    ambiguous: &StringChunked,
+) -> PolarsResult<DatetimeChunked> {
+    let n = ca.len();
+
+    // For each argument, we must check if:
+    // 1. No value was supplied (None)       --> Use existing year from Series
+    // 2. Value was supplied and is a Scalar --> Create full Series of value
+    // 3. Value was supplied and is Series   --> Update all elements with the non-null values
+    let year = if year.len() == 1 {
+        if let Some(value) = year.get(0) {
+            &Int32Chunked::full("".into(), value, n)
+        } else {
+            &ca.year()
+        }
+    } else {
+        &year.zip_with(&year.is_not_null(), &ca.year())?
+    };
+    let month = if month.len() == 1 {
+        if let Some(value) = month.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.month()
+        }
+    } else {
+        &month.zip_with(&month.is_not_null(), &ca.month())?
+    };
+    let day = if day.len() == 1 {
+        if let Some(value) = day.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.day()
+        }
+    } else {
+        &day.zip_with(&day.is_not_null(), &ca.day())?
+    };
+    let hour = if hour.len() == 1 {
+        if let Some(value) = hour.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.hour()
+        }
+    } else {
+        &hour.zip_with(&hour.is_not_null(), &ca.hour())?
+    };
+    let minute = if minute.len() == 1 {
+        if let Some(value) = minute.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.minute()
+        }
+    } else {
+        &minute.zip_with(&minute.is_not_null(), &ca.minute())?
+    };
+    let second = if second.len() == 1 {
+        if let Some(value) = second.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.second()
+        }
+    } else {
+        &second.zip_with(&second.is_not_null(), &ca.second())?
+    };
+    let microsecond = if microsecond.len() == 1 {
+        if let Some(value) = microsecond.get(0) {
+            &Int32Chunked::full("".into(), value, n)
+        } else {
+            &(ca.nanosecond() / 1000)
```

Comment:
are you sure this is right?

Reply:
@MarcoGorelli you're right, I should look into using `and_hms_nano_opt` for the datetime construction.

Reply:
How do we want to do this? Have `microseconds` and `nanoseconds` as mutually exclusive input arguments? Or perhaps `subseconds` which is dependent on the time unit? That's a bit less obvious to the end-user that doesn't supply a time unit in the construction.

Reply:
I think the original nanoseconds need to be preserved, there doesn't necessarily need to be a `nanoseconds` argument (at least, not in this PR, it could be a separate topic for add one here and in `datetime`)

Reply:
As in, the user-facing API could be kept as-is, but `DatetimeChunked::new_from_parts` takes `nanosecond` as argument instead of `microsecond`

then here you make a `nanosecond` variable and pass that

Reply:
@MarcoGorelli the only issue I'm running into is with nanosecond Series, trying to replace values that are outside of the allowable range. Should I just let the panic happen?

This issue exists elsewhere in the codeback, e.g. here:

```python
from datetime import datetime
pl.Series(datetime(1500, 1, 1), dtype=pl.Datetime("ns"))
shape: (1,)
Series: '' [datetime[ns]]
[
        2084-07-20 23:34:33.709551616
]
```
Note the wrong datetime. Much of the construction that goes through chrono silently fails with outside-of-bounds dates. In the case now, if we try to replace outside of the allowable range, it'll panic:

```python
s = pl.Series([datetime(2000, 1, 1)], dtype=pl.Datetime("ns"))
s.dt.replace(year=1900)  # ok, works
s.dt.replace(year=1500)  # panic
```
Not sure whether we should just allow this as a limitation. Catching the error and raising may slow things down a bit.

Reply:
that's fine, that should error (there's a separate issue about making it error instead of panic), if someone has `'ns'` time unit they need to stay within the bounds of that time unit

Reply:
@MarcoGorelli I think in that case the new version is ready for review. I've added in two additional tests to ensure that 1) nanoseconds are preserved if microseconds are not provided, and 2) that time zones & time units are preserved.

Reply:
thanks - could you rebase or merge please?

Reply:
@MarcoGorelli done.